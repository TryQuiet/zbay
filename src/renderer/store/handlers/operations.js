import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import { createAction, handleActions } from 'redux-actions'
import * as R from 'ramda'
import logsHandlers from '../../store/handlers/logs'

import { getClient } from '../../zcash'
import { actionTypes } from '../../../shared/static'

const oneOf = (...arr) => val => R.includes(val, arr)

export const isFinished = oneOf('success', 'cancelled', 'failed')

const POLLING_OFFSET = 15000

export const initialState = Immutable.Map()

export const ZcashError = Immutable.Record({
  code: null,
  message: ''
}, 'ZcashError')

export const ShieldBalanceOp = Immutable.Record({
  amount: new BigNumber(0),
  from: '',
  to: ''
})

export const PendingMessageOp = Immutable.Record({
  message: Immutable.Map(),
  channelId: ''
})

export const PendingDirectMessageOp = Immutable.Record({
  message: Immutable.Map(),
  recipientAddress: '',
  recipientUsername: '',
  offerId: ''
})

export const operationTypes = {
  shieldBalance: 'shieldBalance',
  pendingMessage: 'pendingMessage',
  pendingDirectMessage: 'pendingDirectMessage',
  pendingPlainTransfer: 'pendingPlainTransfer'
}

export const Operation = Immutable.Record({
  opId: '',
  type: '',
  meta: {},
  txId: null,
  error: null,
  status: 'pending'
}, 'Operation')

const addOperation = createAction(actionTypes.ADD_PENDING_OPERATION)
const resolveOperation = createAction(actionTypes.RESOLVE_PENDING_OPERATION)
const removeOperation = createAction(actionTypes.REMOVE_PENDING_OPERATION)

export const actions = {
  addOperation,
  resolveOperation,
  removeOperation
}

const observeOperation = ({ opId, type, meta, checkConfirmationNumber }) => async (dispatch, getState) => {
  dispatch(addOperation({ opId, type, meta }))

  const subscribe = async (callback) => {
    async function poll () {
      const {
        status,
        result: { txid: txId } = {},
        error = null
      } = await getClient().operations.getStatus(opId) || {}
      if (isFinished(status)) {
        return callback(error, { status, txId })
      } else {
        setTimeout(poll, POLLING_OFFSET)
      }
    }
    return poll()
  }

  return subscribe((error, { status, txId }) => {
    dispatch(resolveOperation({ opId, status, txId, error }))
    dispatch(logsHandlers.epics.saveLogs({ type: 'TRANSACTION', payload: txId }))
    if (checkConfirmationNumber && !error) {
      checkConfirmationNumber({ opId, status, txId, getState, dispatch })
    }
  })
}

export const epics = {
  observeOperation
}

export const reducer = handleActions({
  [addOperation]: (state, { payload: { type, meta, opId } }) => state.set(
    opId,
    Operation({
      opId,
      type,
      meta
    })
  ),
  [resolveOperation]: (state, { payload: { opId, status, txId, error } }) => state.update(
    opId,
    m => m.merge({
      status,
      txId,
      error: error ? ZcashError(error) : null
    })
  ),
  [removeOperation]: (state, { payload: id }) => {
    if (state.has(id)) {
      return state.delete(id)
    }
    // It may be a txId
    return state.filter(pm => pm.txId !== id)
  }
}, initialState)

export default {
  actions,
  epics,
  reducer
}
