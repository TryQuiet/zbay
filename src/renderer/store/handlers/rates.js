import Immutable from 'immutable'
import { handleActions, createAction } from 'redux-actions'

import { actionTypes, PRICE_ORACLE_PUB_KEY } from '../../../shared/static'
import channelsSelectors from '../selectors/channels'
import { getClient } from '../../zcash'
import appSelectors from '../selectors/app'
import appHandlers from './app'
import txnTimestampsHandlers from '../handlers/txnTimestamps'
import txnTimestampsSelector from '../selectors/txnTimestamps'
import { getPublicKeysFromSignature } from '../../zbay/messages'
import { trimNull } from '../../zbay/transit'
import { getVault } from '../../vault'
import electronStore from '../../../shared/electronStore'

export const RatesState = Immutable.Record(
  {
    usd: '0',
    zec: '1'
  },
  'RatesState'
)
export const initialState = RatesState({
  usd: '70.45230379033394',
  zec: '1'
})
export const setPriceUsd = createAction(actionTypes.SET_PRICE_USD)
export const actions = {
  setPriceUsd
}
export const setInitialPrice = () => async (dispatch, getState) => {
  try {
    const price = electronStore.get('rates.usd')
    if (price) {
      dispatch(setPriceUsd({ priceUsd: price }))
    }
  } catch (err) {
    console.log(err)
  }
}
export const fetchPrices = () => async (dispatch, getState) => {
  try {
    const channel = channelsSelectors.priceOracleChannel(getState())
    let txnTimestamps = txnTimestampsSelector.tnxTimestamps(getState())
    const transfers = await getClient().payment.received(channel.get('address'))
    if (
      transfers.length ===
      appSelectors.transfers(getState()).get(channel.get('address'))
    ) {
      return
    } else {
      dispatch(
        appHandlers.actions.setTransfers({
          id: channel.get('address'),
          value: transfers.length
        })
      )
    }
    for (const key in transfers) {
      const transfer = transfers[key]
      if (!txnTimestamps.get(transfer.txid)) {
        const result = await getClient().confirmations.getResult(transfer.txid)
        await getVault().transactionsTimestamps.addTransaction(
          transfer.txid,
          result.time
        )
        await dispatch(
          txnTimestampsHandlers.actions.addTxnTimestamp({
            tnxs: { [transfer.txid]: result.time.toString() }
          })
        )
      }
    }
    txnTimestamps = txnTimestampsSelector.tnxTimestamps(getState())
    const sortedTransfers = transfers.sort(
      (b, a) => txnTimestamps.get(a.txid) - txnTimestamps.get(b.txid)
    )
    for (const msg of sortedTransfers) {
      try {
        const memo = trimNull(Buffer.from(msg.memo, 'hex').toString())
        const price = trimNull(memo.substring(129))
        const pkey = getPublicKeysFromSignature({
          message: price,
          signature: Buffer.from(memo.substring(0, 128), 'hex'),
          r: parseInt(memo.substring(128, 129))
        })
        if (pkey.toString('hex') !== PRICE_ORACLE_PUB_KEY) {
          continue
        }
        dispatch(setPriceUsd({ priceUsd: price }))
        electronStore.set('rates.usd', price)
        break
      } catch (err) {
        continue
      }
    }
  } catch (err) {
    console.warn(err)
  }
}
export const fetchPriceForTime = time => async (dispatch, getState) => {
  try {
    const channel = channelsSelectors.priceOracleChannel(getState())
    let txnTimestamps = txnTimestampsSelector.tnxTimestamps(getState())
    const transfers = await getClient().payment.received(channel.get('address'))
    for (const key in transfers) {
      const transfer = transfers[key]
      if (!txnTimestamps.get(transfer.txid)) {
        const result = await getClient().confirmations.getResult(transfer.txid)
        await getVault().transactionsTimestamps.addTransaction(
          transfer.txid,
          result.time
        )
        await dispatch(
          txnTimestampsHandlers.actions.addTxnTimestamp({
            tnxs: { [transfer.txid]: result.time.toString() }
          })
        )
      }
    }
    txnTimestamps = txnTimestampsSelector.tnxTimestamps(getState())
    const closesTransaction = transfers
      .map(txn => txn.txid)
      .reduce((prev, curr) => {
        return Math.abs(parseInt(txnTimestamps.get(curr)) - time) <
          Math.abs(parseInt(txnTimestamps.get(prev)) - time)
          ? curr
          : prev
      })
    const txn = transfers.find(txn => txn.txid === closesTransaction)
    const txnMemo = trimNull(Buffer.from(txn.memo, 'hex').toString())
    const price = trimNull(txnMemo.substring(129))
    return parseFloat(price)
  } catch (err) {
    console.warn(err)
  }
}
export const epics = {
  fetchPrices,
  fetchPriceForTime,
  setInitialPrice
}

export const reducer = handleActions(
  {
    [setPriceUsd]: (state, { payload: { priceUsd } }) =>
      state.set('usd', priceUsd)
  },
  initialState
)

export default {
  reducer,
  actions,
  epics
}
