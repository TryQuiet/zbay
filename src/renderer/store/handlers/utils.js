import Immutable from 'immutable'

import { notifierAction } from '../../components/ui/DismissSnackbarAction'

export const typePending = name => `${name}_PENDING`
export const typeFulfilled = name => `${name}_FULFILLED`
export const typeRejected = name => `${name}_REJECTED`

export const errorNotification = ({ message, options }) => ({
  message,
  options: {
    persist: false,
    variant: 'error',
    action: notifierAction,
    ...options
  }
})

export const successNotification = ({ message, options }) => ({
  message,
  options: {
    variant: 'success',
    ...options
  }
})

export const LoaderState = Immutable.Record({
  loading: false,
  message: ''
})

export const FetchingState = Immutable.Record({
  sizeLeft: 0,
  part: '',
  fetchingStatus: '',
  fetchingSpeed: null,
  fetchingEndTime: {
    hours: null,
    minutes: null,
    seconds: null
  },
  isFetching: false,
  rescanningProgress: 0,
  isRescanningMonitorStarted: false,
  isRescanningInitialized: false,
  guideStatus: false,
  currentSlide: 0
})

export default {
  typePending,
  typeFulfilled,
  typeRejected
}
