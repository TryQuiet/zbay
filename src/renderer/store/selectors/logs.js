import { createSelector } from 'reselect'

const store = s => s

export const logs = createSelector(store, state => state.get('logsData'))

const applicationLogs = createSelector(logs, a => a.applicationLogs)
const transactionsLogs = createSelector(logs, a => a.transactionLogs)
const nodeLogs = createSelector(logs, a => a.nodeLogs)
const islogsFileLoaded = createSelector(logs, a => a.islogsFileLoaded)
const isLogWindowOpened = createSelector(logs, a => a.isLogWindowOpened)

export default {
  applicationLogs,
  transactionsLogs,
  nodeLogs,
  islogsFileLoaded,
  isLogWindowOpened
}
