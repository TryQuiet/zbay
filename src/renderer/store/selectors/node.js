import { DateTime, Interval } from 'luxon'
import { createSelector } from 'reselect'
import BigNumber from 'bignumber.js'

const store = s => s

export const node = createSelector(store, state => state.get('node'))

const currentBlock = createSelector(node, n => n.currentBlock)
const latestBlock = createSelector(node, n => n.latestBlock)
const connections = createSelector(node, n => n.connections)
const status = createSelector(node, n => n.status)
const uptime = createSelector(
  node,
  n => Interval
    .fromDateTimes(DateTime.fromISO(n.startedAt), DateTime.utc())
    .toDuration(['days', 'hours', 'minutes', 'seconds'])
    .normalize()
    .toObject()
)
const percentSynced = createSelector(
  [currentBlock, latestBlock],
  (current, latest) => {
    if (!latest.isZero()) {
      return current.dividedBy(latest).multipliedBy(100).toFixed(0, BigNumber.ROUND_DOWN)
    }
    return null
  }
)
const network = createSelector(node, n => {
  if (parseInt(process.env.ZBAY_IS_TESTNET) === 1) {
    return 'testnet'
  } else {
    return 'mainnet'
  }
})

const isConnected = createSelector(status, s => ['healthy', 'syncing'].includes(s))

const loader = createSelector(node, n => n.bootstrapLoader)
const fetching = createSelector(node, n => n.fetchingStatus)
const bootstrapping = createSelector(loader, n => n.loading)
const bootstrappingMessage = createSelector(loader, n => n.message)
const fetchingPart = createSelector(fetching, n => n.part)
const fetchingSize = createSelector(fetching, n => n.sizeLeft)
const fetchingStatus = createSelector(fetching, n => n.fetchingStatus)
const fetchingSpeed = createSelector(fetching, n => n.fetchingSpeed)
const fetchingEndTime = createSelector(fetching, n => n.fetchingEndTime)
const rescanningProgress = createSelector(fetching, n => n.rescanningProgress)
const isFetching = createSelector(fetching, n => n.isFetching)
const isRescanningMonitorStarted = createSelector(fetching, n => n.isRescanningMonitorStarted)
const isRescanningInitialized = createSelector(fetching, n => n.isRescanningInitialized)
const guideStatus = createSelector(fetching, n => n.guideStatus)

export default {
  node,
  currentBlock,
  latestBlock,
  status,
  uptime,
  connections,
  network,
  percentSynced,
  isConnected,
  bootstrapping,
  bootstrappingMessage,
  fetchingPart,
  fetchingSize,
  fetchingStatus,
  fetchingSpeed,
  isFetching,
  fetchingEndTime,
  rescanningProgress,
  isRescanningMonitorStarted,
  isRescanningInitialized,
  guideStatus
}
