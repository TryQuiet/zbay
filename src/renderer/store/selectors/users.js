import { createSelector } from 'reselect'
import identitySelectors from './identity'
const store = s => s

const users = createSelector(store, state => {
  return state.get('users')
})

const isRegisteredUsername = nickname => createSelector(users, (users) => {
  return users.toList().map(user => user.get('nickname')).includes(nickname)
})

const registeredUser = signerPubKey =>
  createSelector(users, users => users.get(signerPubKey))

const myUser = createSelector(
  users,
  identitySelectors.signerPubKey,
  (users, signerPubKey) => {
    return (
      users.get(signerPubKey) || {
        firstName: '',
        lastName: '',
        nickname: 'Anon' + signerPubKey.substring(0, 15),
        address: '',
        createdAt: 0
      }
    )
  }
)

export default {
  users,
  myUser,
  registeredUser,
  isRegisteredUsername
}
