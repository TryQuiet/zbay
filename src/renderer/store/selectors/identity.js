import BigNumber from 'bignumber.js'
import { createSelector } from 'reselect'
import * as R from 'ramda'

import { rate } from './rates'
import operationsSelectors from './operations'
import { operationTypes } from '../handlers/operations'

const store = s => s

const identity = createSelector(store, state => state.get('identity'))

const data = createSelector(identity, i => i.data)

const balance = currency =>
  createSelector(data, rate(currency), (d, rate) =>
    rate.times(new BigNumber(d.balance || 0))
  )

const lockedBalance = currency =>
  createSelector(data, rate(currency), (d, rate) =>
    rate.times(new BigNumber(d.lockedBalance || 0))
  )

const id = createSelector(data, d => d.id)
const name = createSelector(data, d => d.name)

const donationAllow = createSelector(data, d => d.donationAllow)
const freeUtxos = createSelector(data, d => d.freeUtxos)
const shieldingTax = createSelector(data, d => d.shieldingTax)
const donationAddress = createSelector(data, d => d.donationAddress)

const donation = createSelector(data, d => ({
  allow: d.donationAllow,
  address: d.donationAddress
}))
const signerPrivKey = createSelector(data, d => d.signerPrivKey)
const signerPubKey = createSelector(data, d => d.signerPubKey)

const address = createSelector(data, d => d.address)
const transparentAddress = createSelector(data, d => d.transparentAddress)
const topAddress = createSelector(
  data,
  d => d.addresses.get(0) || d.transparentAddress
)
const addresses = createSelector(data, d => d.addresses)
const topShieldedAddress = createSelector(
  data,
  d => d.shieldedAddresses.get(0) || d.address
)
const shieldedAddresses = createSelector(data, d => d.shieldedAddresses)

const _isShieldOperation = R.curry((identity, op) => {
  return (
    op.type === operationTypes.shieldBalance &&
    op.meta.from === identity.transparentAddress &&
    op.meta.to === identity.address
  )
})

const transparentBalance = createSelector(
  data,
  operationsSelectors.operations,
  (d, ops) =>
    R.compose(
      R.reduce(
        (acc, [key, val]) => acc.plus(val.meta.amount),
        new BigNumber(0)
      ),
      o => o.filter(_isShieldOperation(d))
    )(ops)
)

const loader = createSelector(identity, i => i.loader)

const shippingData = createSelector(data, d => d.shippingData)

export default {
  id,
  name,
  data,
  identity,
  address,
  transparentAddress,
  transparentBalance,
  lockedBalance,
  balance,
  loader,
  shippingData,
  signerPrivKey,
  signerPubKey,
  donationAllow,
  donationAddress,
  donation,
  shieldingTax,
  freeUtxos,
  topAddress,
  topShieldedAddress,
  addresses,
  shieldedAddresses
}
