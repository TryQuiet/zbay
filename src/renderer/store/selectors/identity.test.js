/* eslint import/first: 0 */
jest.mock('../../vault')
import Immutable from 'immutable'
import BigNumber from 'bignumber.js'

import create from '../create'
import { IdentityState, Identity, ShippingData } from '../handlers/identity'
import { RatesState } from '../handlers/rates'
import { Operation, operationTypes, ShieldBalanceOp } from '../handlers/operations'
import { LoaderState } from '../handlers/utils'
import selectors from './identity'

describe('identity selectors', () => {
  const transparentAddress = 't14oHp2v54vfmdgQ3v3SNuQga8JKHTNi2a1'
  const address = 'zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9sly'
  const signerPrivKey = Buffer.alloc(32)
  const signerPubKey = Buffer.alloc(32)
  const shippingData = {
    firstName: 'Saturn',
    lastName: 'the Planet',
    street: 'Coders Dv',
    country: 'Poland',
    region: 'Malopolska',
    city: 'Krakow',
    postalCode: '1337-455'
  }
  let store = null
  beforeEach(() => {
    store = create({
      initialState: Immutable.Map({
        identity: IdentityState({
          data: Identity({
            address,
            transparentAddress,
            name: 'Saturn',
            transparentBalance: '12.123456',
            balance: '33.583004',
            lockedBalance: '12.583004',
            shippingData: ShippingData(shippingData),
            signerPrivKey,
            signerPubKey,
            donationAllow: 'false',
            donationAddress: 'test'
          }),
          loader: LoaderState({
            message: 'Test loading message',
            loading: true
          })
        }),
        rates: RatesState({
          zec: '1',
          usd: '2'
        }),
        operations: Immutable.Map({
          'test-operation-id': Operation({
            opId: 'test-operation-id',
            type: operationTypes.shieldBalance,
            meta: ShieldBalanceOp({
              amount: new BigNumber('0.1234'),
              from: transparentAddress,
              to: address
            })
          }),
          'test-operation-id-1': Operation({
            opId: 'test-operation-id-1',
            type: 'not-shield-operation',
            meta: { name: 'test' }
          }),
          'test-operation-id-2': Operation({
            opId: 'test-operation-id-2',
            type: operationTypes.shieldBalance,
            meta: ShieldBalanceOp({
              amount: new BigNumber('0.2345'),
              from: transparentAddress,
              to: address
            })
          })
        })
      })
    })
    jest.clearAllMocks()
  })

  it('identity', () => {
    expect(selectors.identity(store.getState())).toMatchSnapshot()
  })

  each(['usd', 'zec']).test('balance for %s', currency => {
    expect(selectors.balance(currency)(store.getState())).toMatchSnapshot()
  })

  it('address', () => {
    expect(selectors.address(store.getState())).toMatchSnapshot()
  })
  it('topShieldedAddress', () => {
    expect(selectors.topShieldedAddress(store.getState())).toMatchSnapshot()
  })
  it('topAddress', () => {
    expect(selectors.topAddress(store.getState())).toMatchSnapshot()
  })

  it('transparentAddress', () => {
    expect(selectors.transparentAddress(store.getState())).toMatchSnapshot()
  })

  it('transparentBalance', () => {
    expect(selectors.transparentBalance(store.getState())).toMatchSnapshot()
  })

  each(['usd', 'zec']).test('lockedBalance for %s', currency => {
    expect(selectors.lockedBalance(currency)(store.getState())).toMatchSnapshot()
  })

  it('data', () => {
    expect(selectors.data(store.getState())).toMatchSnapshot()
  })

  it('loader', () => {
    expect(selectors.loader(store.getState())).toMatchSnapshot()
  })

  it('signerPrivKey', () => {
    expect(selectors.signerPrivKey(store.getState())).toMatchSnapshot()
  })

  it('signerPubKey', () => {
    expect(selectors.signerPubKey(store.getState())).toMatchSnapshot()
  })

  it('shippingData', () => {
    expect(selectors.shippingData(store.getState())).toMatchSnapshot()
  })

  it('donationAllow', () => {
    expect(selectors.donationAllow(store.getState())).toMatchSnapshot()
  })

  it('donationAddress', () => {
    expect(selectors.donationAddress(store.getState())).toMatchSnapshot()
  })
  it('donation', () => {
    expect(selectors.donation(store.getState())).toMatchSnapshot()
  })
})
