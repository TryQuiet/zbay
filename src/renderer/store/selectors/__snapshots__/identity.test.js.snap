// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`identity selectors address 1`] = `"zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9sly"`;

exports[`identity selectors balance for usd 1`] = `"67.166008"`;

exports[`identity selectors balance for zec 1`] = `"33.583004"`;

exports[`identity selectors data 1`] = `
Immutable.Identity {
  "id": null,
  "address": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9sly",
  "transparentAddress": "t14oHp2v54vfmdgQ3v3SNuQga8JKHTNi2a1",
  "signerPrivKey": Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  "signerPubKey": Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  "name": "Saturn",
  "shippingData": Immutable.ShippingData {
    "firstName": "Saturn",
    "lastName": "the Planet",
    "street": "Coders Dv",
    "country": "Poland",
    "region": "Malopolska",
    "city": "Krakow",
    "postalCode": "1337-455",
  },
  "balance": "33.583004",
  "lockedBalance": "12.583004",
  "donationAllow": "false",
  "shieldingTax": true,
  "donationAddress": "test",
  "freeUtxos": 0,
  "addresses": Immutable.List [],
  "shieldedAddresses": Immutable.List [],
}
`;

exports[`identity selectors donation 1`] = `
Object {
  "address": "test",
  "allow": "false",
}
`;

exports[`identity selectors donationAddress 1`] = `"test"`;

exports[`identity selectors donationAllow 1`] = `"false"`;

exports[`identity selectors identity 1`] = `
Immutable.IdentityState {
  "data": Immutable.Identity {
    "id": null,
    "address": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9sly",
    "transparentAddress": "t14oHp2v54vfmdgQ3v3SNuQga8JKHTNi2a1",
    "signerPrivKey": Object {
      "data": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "type": "Buffer",
    },
    "signerPubKey": Object {
      "data": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "type": "Buffer",
    },
    "name": "Saturn",
    "shippingData": Immutable.ShippingData {
      "firstName": "Saturn",
      "lastName": "the Planet",
      "street": "Coders Dv",
      "country": "Poland",
      "region": "Malopolska",
      "city": "Krakow",
      "postalCode": "1337-455",
    },
    "balance": "33.583004",
    "lockedBalance": "12.583004",
    "donationAllow": "false",
    "shieldingTax": true,
    "donationAddress": "test",
    "freeUtxos": 0,
    "addresses": Immutable.List [],
    "shieldedAddresses": Immutable.List [],
  },
  "fetchingBalance": false,
  "loader": Immutable.Record {
    "loading": true,
    "message": "Test loading message",
  },
  "errors": "",
}
`;

exports[`identity selectors loader 1`] = `
Immutable.Record {
  "loading": true,
  "message": "Test loading message",
}
`;

exports[`identity selectors lockedBalance for usd 1`] = `"25.166008"`;

exports[`identity selectors lockedBalance for zec 1`] = `"12.583004"`;

exports[`identity selectors shippingData 1`] = `
Immutable.ShippingData {
  "firstName": "Saturn",
  "lastName": "the Planet",
  "street": "Coders Dv",
  "country": "Poland",
  "region": "Malopolska",
  "city": "Krakow",
  "postalCode": "1337-455",
}
`;

exports[`identity selectors signerPrivKey 1`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`identity selectors signerPubKey 1`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`identity selectors topAddress 1`] = `"t14oHp2v54vfmdgQ3v3SNuQga8JKHTNi2a1"`;

exports[`identity selectors topShieldedAddress 1`] = `"zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9sly"`;

exports[`identity selectors transparentAddress 1`] = `"t14oHp2v54vfmdgQ3v3SNuQga8JKHTNi2a1"`;

exports[`identity selectors transparentBalance 1`] = `"0.3579"`;
